### PR TITLE
Distributed Tracing: Fixes traceid null exception issue

### LINKED ISSUE #4080: OpenTelemetry - NullReferenceException
Using SDK 3.35.1-preview, we have 2 reports of NullReferenceException:

```
System.NullReferenceException
  HResult=0x80004003
  Message=Object reference not set to an instance of an object.
  Source=Microsoft.Azure.Cosmos.Client
  StackTrace:
   at Microsoft.Azure.Cosmos.Tracing.TraceWriter.WriteTraceDatum(IJsonWriter writer, Object value)
   at Microsoft.Azure.Cosmos.Tracing.TraceWriter.TraceJsonWriter.WriteTrace(IJsonWriter writer, ITrace trace, Boolean isRootTrace)
   at Microsoft.Azure.Cosmos.Tracing.TraceWriter.WriteTrace(IJsonWriter writer, ITrace trace)
   at Microsoft.Azure.Cosmos.Diagnostics.CosmosTraceDiagnostics.WriteTraceToJsonWriter(JsonSerializationFormat jsonSerializationFormat)
   at Microsoft.Azure.Cosmos.Diagnostics.CosmosTraceDiagnostics.ToJsonString()
   at Microsoft.Azure.Cosmos.Diagnostics.CosmosTraceDiagnostics.ToString()
   at Microsoft.Azure.Cosmos.CosmosException.GetMessageHelper(HttpStatusCode statusCode, Headers headers, String responseBody, CosmosDiagnostics diagnostics)raceWriterTrac
   at Microsoft.Azure.Cosmos.CosmosException.<>c__DisplayClass2_0.<.ctor>b__0()
   at System.Lazy`1.ViaFactory(LazyThreadSafetyMode mode)
   at System.Lazy`1.ExecutionAndPublication(LazyHelper executionAndPublication, Boolean useDefaultConstructor)
   at System.Lazy`1.CreateValue()
   at Microsoft.Azure.Cosmos.CosmosException.get_Message()
   at Microsoft.Azure.Cosmos.CosmosException.RecordOtelAttributes(CosmosException exception, DiagnosticScope scope)
   at Microsoft.Azure.Cosmos.Telemetry.OpenTelemetryCoreRecorder.<>c.<.cctor>b__22_3(Exception exception, DiagnosticScope scope)
   at Microsoft.Azure.Cosmos.Telemetry.OpenTelemetryCoreRecorder.IsExceptionRegistered(Exception exception, DiagnosticScope scope)
   at Microsoft.Azure.Cosmos.Telemetry.OpenTelemetryCoreRecorder.MarkFailed(Exception exception)
   at Microsoft.Azure.Cosmos.ClientContextCore.<RunWithDiagnosticsHelperAsync>d__40`1.MoveNext()
   at Microsoft.Azure.Cosmos.ClientContextCore.<OperationHelperWithRootTraceAsync>d__30`1.MoveNext()
   at .....

 

  This exception was originally thrown at this call stack:
    Microsoft.Azure.Cosmos.Tracing.TraceWriter.WriteTraceDatum(Microsoft.Azure.Cosmos.Json.IJsonWriter, object)
    Microsoft.Azure.Cosmos.Tracing.TraceWriter.TraceJsonWriter.WriteTrace(Microsoft.Azure.Cosmos.Json.IJsonWriter, Microsoft.Azure.Cosmos.Tracing.ITrace, bool)
    Microsoft.Azure.Cosmos.Tracing.TraceWriter.WriteTrace(Microsoft.Azure.Cosmos.Json.IJsonWriter, Microsoft.Azure.Cosmos.Tracing.ITrace)
    Microsoft.Azure.Cosmos.Diagnostics.CosmosTraceDiagnostics.WriteTraceToJsonWriter(Microsoft.Azure.Cosmos.Json.JsonSerializationFormat)
    Microsoft.Azure.Cosmos.Diagnostics.CosmosTraceDiagnostics.ToJsonString()
    Microsoft.Azure.Cosmos.Diagnostics.CosmosTraceDiagnostics.ToString()
    Microsoft.Azure.Cosmos.CosmosException.GetMessageHelper(System.Net.HttpStatusCode, Microsoft.Azure.Cosmos.Headers, string, Microsoft.Azure.Cosmos.CosmosDiagnostics)
```

Another example:

```
at Microsoft.Azure.Cosmos.Tracing.TraceWriter.WriteTraceDatum(IJsonWriter writer, Object value)
   at Microsoft.Azure.Cosmos.Tracing.TraceWriter.TraceJsonWriter.WriteTrace(IJsonWriter writer, ITrace trace, Boolean isRootTrace)
   at Microsoft.Azure.Cosmos.Tracing.TraceWriter.WriteTrace(IJsonWriter writer, ITrace trace)
   at Microsoft.Azure.Cosmos.Diagnostics.CosmosTraceDiagnostics.WriteTraceToJsonWriter(JsonSerializationFormat jsonSerializationFormat)
   at Microsoft.Azure.Cosmos.Diagnostics.CosmosTraceDiagnostics.ToJsonString()
   at Microsoft.Azure.Cosmos.Diagnostics.CosmosTraceDiagnostics.ToString()
   at Microsoft.Azure.Cosmos.CosmosException.GetMessageHelper(HttpStatusCode statusCode, Headers headers, String responseBody, CosmosDiagnostics diagnostics)
   at Microsoft.Azure.Cosmos.CosmosException.<>c__DisplayClass2_0.<.ctor>b__0()
   at System.Lazy`1.ViaFactory(LazyThreadSafetyMode mode)
   at System.Lazy`1.ExecutionAndPublication(LazyHelper executionAndPublication, Boolean useDefaultConstructor)
   at System.Lazy`1.CreateValue()
   at Microsoft.Azure.Cosmos.CosmosException.get_Message()
   at Microsoft.Azure.Cosmos.CosmosException.RecordOtelAttributes(CosmosException exception, DiagnosticScope scope)
   at Microsoft.Azure.Cosmos.Telemetry.OpenTelemetryCoreRecorder.<>c.<.cctor>b__22_3(Exception exception, DiagnosticScope scope)
   at Microsoft.Azure.Cosmos.Telemetry.OpenTelemetryCoreRecorder.IsExceptionRegistered(Exception exception, DiagnosticScope scope)
   at Microsoft.Azure.Cosmos.Telemetry.OpenTelemetryCoreRecorder.MarkFailed(Exception exception)
   at Microsoft.Azure.Cosmos.ClientContextCore.<RunWithDiagnosticsHelperAsync>d__40`1.MoveNext()
   at Microsoft.Azure.Cosmos.ClientContextCore.<OperationHelperWithRootTraceAsync>d__30`1.MoveNext()
   at CFPullModelAllVersionsAndDeletesMode.Program.<ReadChangeFeed>d__16.MoveNext() in C:\Users\rodimiroc\OneDrive - Microsoft\Documents\Downloads\FFCF-solution2\cosmosdb-changefeed\Program.cs:line 211
```

The problem happens consistently. The way to repro it is generate any error condition, for example, attempt to Create a document that contains a `ttl` property with `null` value.

```
public class ReproItem
{
    public string id {get;set;} = "id";

    public int? ttl {get;set}
}
```

Which generates a CosmosException, catch the CosmosException and try to print the Diagnostics.ToString()

Workaround currently is to disable Open Telemetry integration: `CosmosClientOptions.IsDistributedTracingEnabled = false`;

### PR DESCRIPTION
**Pull request by aavasthy**
# Pull Request Template

## Description

When using SDK preview version 3.35.1-preview and above, the diagnostic logs generation(Diagnostics.ToString()) throws null reference exception when CosmosClientOptions IsDistributedTracingEnabled as true and not specifying the source it needs to trace. As part of Distributed Tracing feature, a parent level activity is created every time the feature flag is on. The root cause is when listeners are available but not subscribed to any event then it will lead to scope being enabled but no activity is created.

- Updated the parent level activity creation so as to ensure that an activity is created at parent level everytime the IsDistributedTracingEnabled is on.
- Added a safety check for null when adding TraceId in diagnostics log so that the diagnostic generaton doesn't lead to code breaks. 
## Type of change
- [x] Bug fix (non-breaking change which fixes an issue)

## Closing issues

To automatically close an issue: closes #4080

### CODE REVIEW

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/DistributedTracingOTelTests.cs [hunk 1]
```diff
@@ -19,6 +19,8 @@ namespace Microsoft.Azure.Cosmos
     using System.Diagnostics;
     using Microsoft.Azure.Cosmos.Tracing;
     using System.Net.Http;
+    using System.ComponentModel;
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/DistributedTracingOTelTests.cs:22]: Is this using needed?

File: Microsoft.Azure.Cosmos/src/Telemetry/OpenTelemetry/OpenTelemetryRecorderFactory.cs [hunk 2]
```diff
@@ -63,14 +64,27 @@ internal static class OpenTelemetryRecorderFactory
...
+                // If there are no listeners at operation level and network level and no parent activity
+                // then create a dummy parent activity.
+                if (Activity.Current is null)
+                {
+                    openTelemetryRecorder = OpenTelemetryCoreRecorder.CreateParentActivity(operationName);
                 }
 #endif
-                trace.AddDatum("DistributedTraceId", Activity.Current?.TraceId);
+                // Safety check as diagnostic logs should not break the code.
+                if (Activity.Current?.TraceId != null)
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Telemetry/OpenTelemetry/OpenTelemetryRecorderFactory.cs:84]: Since this depends on `Activity.Current` not being `null`, should this be on the `else` of the above `if` which would remove the need of the `?` ?

File: Microsoft.Azure.Cosmos/src/Telemetry/OpenTelemetry/OpenTelemetryRecorderFactory.cs [hunk 3]
```diff
@@ -63,14 +64,27 @@ internal static class OpenTelemetryRecorderFactory
...
+                {
+                    trace.AddDatum("DistributedTraceId", Activity.Current?.TraceId);
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Telemetry/OpenTelemetry/OpenTelemetryRecorderFactory.cs:86]: At this point, nothing is `null`, all 3 items were checked, right?

```suggestion
                    trace.AddDatum("DistributedTraceId", Activity.Current.TraceId);
```

**Review by ealsur — COMMENTED** [review]
What was the line that was generating the NullRef? Please fix the PR title

File: Microsoft.Azure.Cosmos/src/Telemetry/OpenTelemetry/OpenTelemetryRecorderFactory.cs (hunk 2 above)
Comment (aavasthy) [reply Microsoft.Azure.Cosmos/src/Telemetry/OpenTelemetry/OpenTelemetryRecorderFactory.cs:84]: Not including adding traceid in diagnostics in the else part because the if check block for Activity.Current is null also includes creation of activity. Then this activity Id will be added as traceId in diagnostics.

File: Microsoft.Azure.Cosmos/src/Telemetry/OpenTelemetry/OpenTelemetryRecorderFactory.cs (hunk 3 above)
Comment (aavasthy) [reply Microsoft.Azure.Cosmos/src/Telemetry/OpenTelemetry/OpenTelemetryRecorderFactory.cs:86]: removed

**Comment by aavasthy** [comment]
> What was the line that was generating the NullRef? Please fix the PR title

This line was adding traceId as null : trace.AddDatum("DistributedTraceId", Activity.Current.TraceId).
And when doing Diagnostics.ToString() it would throw null exception.

With the initial code, in ideal scenarios when the enableDistributed tracing flag is on and a source is attached then it should create an activity by the time it reaches this line trace.AddDatum("DistributedTraceId", Activity.Current.TraceId).
The issue is being replicated when in certain environments there are subscriptions which leads to scope(operation/network) to get enabled but it doesn't create an activity as there is no source attached. 

So the code is updated to ensure that an activity at parent level is always created.

File: Microsoft.Azure.Cosmos/src/Telemetry/OpenTelemetry/OpenTelemetryRecorderFactory.cs (hunk 2 above)
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/Telemetry/OpenTelemetry/OpenTelemetryRecorderFactory.cs:84]: My point is, you can rewrite this `if` as:

```
if (Activity.Current is null)
{
	openTelemetryRecorder = OpenTelemetryCoreRecorder.CreateParentActivity(operationName);
}
else if (Activity.Current.TraceId != null) 
{
	// at this point Activity.Current cannot be null because if it was, it would have gone into the first id
	trace.AddDatum("DistributedTraceId", Activity.Current.TraceId);
}
```

That way you don't need the `?` in `Activity.Current?.TraceId != null`

Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/Telemetry/OpenTelemetry/OpenTelemetryRecorderFactory.cs:84]: Unless `CreateParentActivity` can create an `Activity.Current`?

File: Microsoft.Azure.Cosmos/src/Telemetry/OpenTelemetry/OpenTelemetryRecorderFactory.cs [hunk 4]
```diff
@@ -48,7 +48,8 @@ internal static class OpenTelemetryRecorderFactory
                 DiagnosticScope scope = LazyOperationScopeFactory.Value.CreateScope(name: operationName,
                                  kind: clientContext.ClientOptions.ConnectionMode == ConnectionMode.Gateway ? DiagnosticScope.ActivityKind.Internal : DiagnosticScope.ActivityKind.Client);
 
-                // Record values only when we have a valid Diagnostic Scope
+                // The scope here checks for listeners at Operation Level.
+                // If there are listeners at operation level then scope is enabled and activity is created.
```
Comment (sourabh1007) [inline Microsoft.Azure.Cosmos/src/Telemetry/OpenTelemetry/OpenTelemetryRecorderFactory.cs:52]: "activity is created" => this is not true. If listeners are there, then scope will be enabled but we don't guarantee that activity will be created.

File: Microsoft.Azure.Cosmos/src/Telemetry/OpenTelemetry/OpenTelemetryRecorderFactory.cs [hunk 5]
```diff
@@ -63,14 +64,27 @@ internal static class OpenTelemetryRecorderFactory
                         config: requestOptions?.CosmosThresholdOptions ?? clientContext.ClientOptions?.CosmosClientTelemetryOptions.CosmosThresholdOptions);
                 }
 #if !INTERNAL
-                else if (Activity.Current is null)
+                // The scope here checks for listeners at Network Level.
+                // If there are listeners at network level and no parent activity created at operation level
+                // then create a network scope that creates a parent activity.
```
Comment (sourabh1007) [inline Microsoft.Azure.Cosmos/src/Telemetry/OpenTelemetry/OpenTelemetryRecorderFactory.cs:69]: in comments, along with "what" mention "why" also. because it should be clear why we decided to create parent activity for network level subscriptions? Why we are making sure that activity is always there?

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/Tracing/CustomOtelExporter.cs [hunk 6]
```diff
@@ -35,7 +35,8 @@ public override ExportResult Export(in Batch<Activity> batch)
 
             foreach (Activity activity in batch)
             {
-                if (string.Equals(activity.Source.Name, $"{OpenTelemetryAttributeKeys.DiagnosticNamespace}.Operation", StringComparison.OrdinalIgnoreCase))
+                if (string.Equals(activity.Source.Name, $"{OpenTelemetryAttributeKeys.DiagnosticNamespace}.Operation", StringComparison.OrdinalIgnoreCase)
+                    || string.Equals(activity.Source.Name, $"{OpenTelemetryAttributeKeys.DiagnosticNamespace}.Request", StringComparison.OrdinalIgnoreCase))
```
Comment (sourabh1007) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/Tracing/CustomOtelExporter.cs:39]: why did you add this OR condition here? here we are validating only operation level activity.

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/Tracing/AssertActivity.cs [hunk 7]
```diff
@@ -83,7 +83,9 @@ public static void AreEqualAcrossListeners()
         {
             Assert.AreEqual(
                 JsonConvert.SerializeObject(CustomListener.CollectedOperationActivities.OrderBy(x => x.Id)),
-                JsonConvert.SerializeObject(CustomOtelExporter.CollectedActivities.OrderBy(x => x.Id)));
```
Comment (sourabh1007) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/Tracing/AssertActivity.cs:86]: Wondering how it was working till now.

Comment (aavasthy) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/Tracing/AssertActivity.cs:86]: I updated ConsoleExporter to take in network request activities as well. That is why I had to add this filter here.
It was working with E2E tests because they were not creating network level activities.

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/Tracing/CustomOtelExporter.cs (hunk 6 above)
Comment (aavasthy) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/Tracing/CustomOtelExporter.cs:39]: I needed to check activities created when subscribed to only network level activities in DistributedOtelTests.

**Comment by kirankumarkolli** [comment]
> activity is created at parent level everytime the IsDistributedTracingEnabled is on

Can you please help explain why is it needed? 
Is an alternative of excludeing DistributedTraceId in traces an acceptable choice?

**Comment by aavasthy** [comment]
> > activity is created at parent level everytime the IsDistributedTracingEnabled is on
> 
> Can you please help explain why is it needed? Is an alternative of excludeing DistributedTraceId in traces an acceptable choice?

If IsDistributedTracingEnabled is on then we always want to return a traceId back to the user which can be used for activity tracking.

File: Microsoft.Azure.Cosmos/src/Telemetry/OpenTelemetry/OpenTelemetryRecorderFactory.cs (hunk 5 above)
Comment (aavasthy) [reply Microsoft.Azure.Cosmos/src/Telemetry/OpenTelemetry/OpenTelemetryRecorderFactory.cs]: updated the comment

File: Microsoft.Azure.Cosmos/src/Telemetry/OpenTelemetry/OpenTelemetryRecorderFactory.cs [hunk 8]
```diff
@@ -48,7 +48,8 @@ internal static class OpenTelemetryRecorderFactory
                 DiagnosticScope scope = LazyOperationScopeFactory.Value.CreateScope(name: operationName,
                                  kind: clientContext.ClientOptions.ConnectionMode == ConnectionMode.Gateway ? DiagnosticScope.ActivityKind.Internal : DiagnosticScope.ActivityKind.Client);
 
-                // Record values only when we have a valid Diagnostic Scope
+                // The scope here checks for listeners at Operation Level.
+                // If there are listeners at operation level then scope is enabled and it tries to create activity.
```
Comment (sourabh1007) [inline Microsoft.Azure.Cosmos/src/Telemetry/OpenTelemetry/OpenTelemetryRecorderFactory.cs:52]: explain more about it, in which cases activity won't be created.

File: Microsoft.Azure.Cosmos/src/Telemetry/OpenTelemetry/OpenTelemetryRecorderFactory.cs [hunk 9]
```diff
@@ -63,14 +64,27 @@ internal static class OpenTelemetryRecorderFactory
...
+                // Need a parent activity at root level so as to group all network activities under it.
+                else
                 {
                     DiagnosticScope requestScope = LazyNetworkScopeFactory.Value.CreateScope(name: operationName);
+                    openTelemetryRecorder = requestScope.IsEnabled ? OpenTelemetryCoreRecorder.CreateNetworkLevelParentActivity(networkScope: requestScope) : openTelemetryRecorder;
+                }
 
-                    openTelemetryRecorder = requestScope.IsEnabled ? OpenTelemetryCoreRecorder.CreateNetworkLevelParentActivity(networkScope: requestScope) : OpenTelemetryCoreRecorder.CreateParentActivity(operationName);
+                // If there are no listeners at operation level and network level and no parent activity created.
+                // Then create a dummy activity as there should be a parent level activity always when Distributed tracing is on.
```
Comment (sourabh1007) [inline Microsoft.Azure.Cosmos/src/Telemetry/OpenTelemetry/OpenTelemetryRecorderFactory.cs:77]: Explain why we are creating dummy activity.

File: Microsoft.Azure.Cosmos/src/Telemetry/OpenTelemetry/OpenTelemetryRecorderFactory.cs [hunk 10]
```diff
@@ -63,14 +64,27 @@ internal static class OpenTelemetryRecorderFactory
                         config: requestOptions?.CosmosThresholdOptions ?? clientContext.ClientOptions?.CosmosClientTelemetryOptions.CosmosThresholdOptions);
                 }
 #if !INTERNAL
-                else if (Activity.Current is null)
+                // The scope here checks for listeners at Network Level.
+                // If there are listeners at network level then scope is enabled and it tries to create activity.
+                // Need a parent activity at root level so as to group all network activities under it.
```
Comment (sourabh1007) [inline Microsoft.Azure.Cosmos/src/Telemetry/OpenTelemetry/OpenTelemetryRecorderFactory.cs:69]: Why we want to group the network level activities?, for better UX on APM tool or to get a common TraceId which will be logged in diagnostic and customer can use it to search on APM tool.

**Review by philipthomas-MSFT — COMMENTED** [review]
From the description, if IsDistributedTracingEnabled is true with no diagnostic source, then this causes of a NullReferenceException? I am not sure where this is in the code. And do we know the underlying reason for there being no diagnostic source?

File: Microsoft.Azure.Cosmos/src/Telemetry/OpenTelemetry/OpenTelemetryRecorderFactory.cs (hunk 2 above)
Comment (aavasthy) [reply Microsoft.Azure.Cosmos/src/Telemetry/OpenTelemetry/OpenTelemetryRecorderFactory.cs:84]: yes CreateParentActivity will create an Activity.Current which needs can be used for setting Activity.Current.TraceId

**Comment by aavasthy** [comment]
> From the description, if IsDistributedTracingEnabled is true with no diagnostic source, then this causes of a NullReferenceException? I am not sure where this is in the code. And do we know the underlying reason for there being no diagnostic source?

We were observing this issue when the end user is setting the IsDistributedTracingEnabled as true and not specifying the source it needs to trace. As part of Distributed Tracing feature, a parent level activity is created every time the feature flag is on. This root activity id is used for tracing purpose.
The OpenTelemetryRecorderFactory code is responsible for creating the parent level activity. With the existing code, it was assumed that whenever diagnostic scope either at operation or network level is enabled then it will lead to activity creation. But the root cause is when listeners are available but not subscribed to any event then it will lead to scope being enabled but no activity is created. That is why the updated change has a separate Activity is null check to ensure activity gets created always.

**Review by ealsur — APPROVED** [review]

File: Microsoft.Azure.Cosmos/src/Telemetry/OpenTelemetry/OpenTelemetryRecorderFactory.cs (hunk 10 above)
Comment (aavasthy) [reply Microsoft.Azure.Cosmos/src/Telemetry/OpenTelemetry/OpenTelemetryRecorderFactory.cs]: updated the comment

File: Microsoft.Azure.Cosmos/src/Telemetry/OpenTelemetry/OpenTelemetryRecorderFactory.cs (hunk 9 above)
Comment (aavasthy) [reply Microsoft.Azure.Cosmos/src/Telemetry/OpenTelemetry/OpenTelemetryRecorderFactory.cs:78]: updated the comment

File: Microsoft.Azure.Cosmos/src/Telemetry/OpenTelemetry/OpenTelemetryRecorderFactory.cs (hunk 8 above)
Comment (aavasthy) [reply Microsoft.Azure.Cosmos/src/Telemetry/OpenTelemetry/OpenTelemetryRecorderFactory.cs:52]: updated the comment

**Review by ealsur — DISMISSED** [review]

**Review by kirankumarkolli — DISMISSED** [review]

File: Microsoft.Azure.Cosmos/src/Telemetry/OpenTelemetry/OpenTelemetryRecorderFactory.cs (hunk 9 above)
Comment (sourabh1007) [reply Microsoft.Azure.Cosmos/src/Telemetry/OpenTelemetry/OpenTelemetryRecorderFactory.cs:78]: may be rephrase it to something like this "Then create a dummy activity as there should be an activity always, to send a traceid to the backend services through context propagation" It would explain, why we are even creating dummy activity if customer is not using this feature at all.

**Review by kirankumarkolli — APPROVED** [review]
